expr.py

```diff
@@ -0,0 +1,133 @@
+    def __rdiv__(self, other): return Expr('/',  other, self)
+    def __rpow__(self, other): return Expr('**', other, self)
+    def __rmod__(self, other): return Expr('%',  other, self)
+    def __rand__(self, other): return Expr('&',  other, self)
+    def __rxor__(self, other): return Expr('^',  other, self)
+    def __ror__(self, other):  return Expr('|',  other, self)
+    def __rrshift__(self, other):   return Expr('>>',  other, self)
+    def __rlshift__(self, other):   return Expr('<<',  other, self)
+    def __rtruediv__(self, other):  return Expr('/',  other, self)
+    def __rfloordiv__(self, other): return Expr('//',  other, self)
+    
+    # Unary operator overloads
+    def __neg__(self):    return Expr('-', self)
+    def __pos__(self):    return Expr('+', self)
+    def __invert__(self): return Expr('~', self)  
+    
+    def __call__(self, *args): 
+        "Call: for example, use f(x) to create Expr('f', x)"
+        return Expr(self.op, *args)
+    
+    def __contains__(self, x): 
+        "'x in exp' tests if x == exp or x in exp.args."
+        return (x == self 
+                or x in self.args 
+                or any(isinstance(a, Expr) and x in a
+                       for a in self.args))
+     
+    def __eq__(self, other):   
+        "'x == y' evaluates to True or False; does not build an Expr."
+        return (isinstance(other, Expr) 
+                and self.op == other.op 
+                and self.args == other.args)
+    
+    def __lt__(self, other):   
+        "'x <= y' evaluates to True or False; does not build an Expr."
+        return (isinstance(other, Expr) 
+                and self.op < other.op 
+                and self.args < other.args)
+    
+    def __hash__(self): return hash(self.op) ^ hash(self.args)
+    
+    def __repr__(self):
+        op   = self.op
+        args = [str(arg) for arg in self.args]
+        if op.isidentifier():       # f(x) or f(x, y)
+            return '{}({})'.format(op, ', '.join(args))
+        elif len(args) == 1:        # -x or -(x + 1)
+            return op + args[0]
+        else:                       # (x - y)
+            opp = (' ' + op + ' ')
+            return '(' + opp.join(args) + ')'
+
+class Symbol(Expr):
+    "An atomic Expr, such as: x = Symbol('x')"
+    
+    def __init__(self, name):
+        self.name = name
+        Expr.__init__(self, name) # Symbols are Exprs
+        
+    def __repr__(self): return self.name
+
+def symbols(names):
+    "Return a list of Symbols; names is a comma/whitespace delimited str."
+    return [Symbol(name) for name in names.replace(',', ' ').split()]
+
+# An 'Expression' is either an 'Expr' or a number:
+Expression = (Expr, int, float, complex)
+
+def arity(expression):
+    "The number of sub-expressions in this expression."
+    if isinstance(expression, Expr):
+        return len(expression.args)
+    else: # expression is a number
+        return 0
+                
+def ex(arg1, arg2=None, arg3=None):
```

Comment:
I like ex() more compared to our old method expr(). This is more flexible as we can give infix and prefix through the same function which we can't do in expr().